lib/matplotlib/tests/test_figure.py

```diff
@@ -923,6 +923,244 @@ def test_animated_with_canvas_change(fig_test, fig_ref):
+        subplot_kw = subfigure_kw or {}
+
+        gs = fig_ref.add_gridspec(1, 2)
+        axA = fig_ref.add_subfigure(gs[0, 0], **subfigure_kw)
+
+        axB = fig_ref.add_subfigure(gs[0, 1], **subfigure_kw)
+
+    @check_figures_equal(extensions=["png"])
+    @pytest.mark.parametrize("multi_value", ['BC', tuple('BC')])
+    def test_per_subfigure_kw(self, fig_test, fig_ref, multi_value):
+        x = 'AB;CD'
+        grid_axes = fig_test.subfigure_mosaic(
+            x,
+            subfigure_kw={'facecolor': 'red'},
+            per_subfigure_kw={
+                'D': {'facecolor': 'blue'},
+                multi_value: {'facecolor': 'green'},
+            }
+        )
+
+        gs = fig_ref.add_gridspec(2, 2)
+        for color, spec in zip(['red', 'green', 'green', 'blue'], gs):
+            fig_ref.add_subfigure(spec, facecolor=color)
+
+    def test_extra_per_subfigure_kw(self):
+        with pytest.raises(
+                ValueError, match=f'The keys {set("B")!r} are in'
+        ):
+            Figure().subfigure_mosaic("A", per_subfigure_kw={"B": {}})
+
+    @check_figures_equal(extensions=["png"])
+    @pytest.mark.parametrize("str_pattern",
+                             ["AAA\nBBB", "\nAAA\nBBB\n", "ABC\nDEF"]
+                             )
+    def test_single_str_input(self, fig_test, fig_ref, str_pattern):
+        grid_axes = fig_test.subfigure_mosaic(str_pattern)
+
+        grid_axes = fig_ref.subfigure_mosaic(
+            [list(ln) for ln in str_pattern.strip().split("\n")]
+        )
+
+    @pytest.mark.parametrize(
+        "x,match",
+        [
+            (
+                    [["A", "."], [".", "A"]],
+                    (
+                            "(?m)we found that the label .A. specifies a "
+                            + "non-rectangular or non-contiguous area."
+                    ),
+            ),
+            (
+                    [["A", "B"], [None, [["A", "B"], ["C", "D"]]]],
+                    "There are duplicate keys .* between the outer layout",
+            ),
+            ("AAA\nc\nBBB", "All of the rows must be the same length"),
+            (
+                    [["A", [["B", "C"], ["D"]]], ["E", "E"]],
+                    "All of the rows must be the same length",
+            ),
+        ],
+    )
+    def test_fail(self, x, match):
+        fig = plt.figure()
+        with pytest.raises(ValueError, match=match):
+            fig.subfigure_mosaic(x)
+
+    @check_figures_equal(extensions=["png"])
+    def test_hashable_keys(self, fig_test, fig_ref):
+        fig_test.subfigure_mosaic([[object(), object()]])
+        fig_ref.subfigure_mosaic([["A", "B"]])
+
+    @pytest.mark.parametrize('str_pattern',
+                             ['abc', 'cab', 'bca', 'cba', 'acb', 'bac'])
+    def test_user_order(self, str_pattern):
+        fig, ax_dict = plt.subfigure_mosaic(str_pattern)
+        assert list(str_pattern) == list(ax_dict)
```

Comment:
In TestSubplotMosaic there's an additional  `assert list(fig.axes) == list(ax_dict.values()) ` statement here and in `test_nested_user_order`. Is there a way to return all subfigures the same way you could return all axes? I tried looking for it but didn't find one, so right now both of these tests only check the dictionary keys being correct.

Reply:
`fig.subfigs`, but it is not recursive.

Reply:
as a side note, I also had to look this up.  My thought process was to assume that if we had a registry, then we probably update it is `FigureBase.add_subfigure` (as that is where we create the new `SubFigure` and that has the line https://github.com/matplotlib/matplotlib/blob/1230846b1b4b1bd442444c437c79c5dfdcc47534/lib/matplotlib/figure.py#L1615